src/SystemView.cpp

```diff
@@ -1069,28 +661,128 @@ void SystemView::DrawGrid()
+		m_lineVerts->Add(vector3f(+m_grid_lines * zoom, 0.0f, z) + vector3f(pos), svColor[GRID]);
 	}
 
 	for (int i = -m_grid_lines; i < m_grid_lines + 1; i++) {
 		float x = float(i) * zoom;
-		m_lineVerts->Add(vector3f(x, 0.0f, -m_grid_lines * zoom) + vector3f(pos), Color::GRAY);
-		m_lineVerts->Add(vector3f(x, 0.0f, +m_grid_lines * zoom) + vector3f(pos), Color::GRAY);
+		m_lineVerts->Add(vector3f(x, 0.0f, -m_grid_lines * zoom) + vector3f(pos), svColor[GRID]);
+		m_lineVerts->Add(vector3f(x, 0.0f, +m_grid_lines * zoom) + vector3f(pos), svColor[GRID]);
 	}
 
 	for (SystemBody *sbody : m_displayed_sbody) {
 		vector3d offset(0.);
 		GetTransformTo(sbody, offset);
-		m_lineVerts->Add(vector3f(pos - offset), Color::GRAY * 0.5);
+		offset *= m_zoom;
+		m_lineVerts->Add(vector3f(pos - offset), svColor[GRID] * 0.5);
 		offset.y = 0.0;
-		m_lineVerts->Add(vector3f(pos - offset), Color::GRAY * 0.5);
+		m_lineVerts->Add(vector3f(pos - offset), svColor[GRID] * 0.5);
 	}
 
 	m_lines.SetData(m_lineVerts->GetNumVerts(), &m_lineVerts->position[0], &m_lineVerts->diffuse[0]);
 	m_lines.Draw(Pi::renderer, m_lineState);
 }
+
+template <typename T>
+void SystemView::AddNotProjected(Projectable::types type, Projectable::bases base, T *ref, const vector3d &worldscaledpos)
+{
+	//project and add
+	Gui::Screen::EnterOrtho();
+	vector3d pos;
+	if (Gui::Screen::Project(worldscaledpos, pos) && pos.z < 1)
+		AddProjected<T>(type, base, ref, pos);
+	Gui::Screen::LeaveOrtho();
+}
+
+template <typename T>
+void SystemView::AddProjected(Projectable::types type, Projectable::bases base, T *ref, vector3d &pos)
+{
+	float scale[2];
+	Gui::Screen::GetCoords2Pixels(scale);
+	Projectable p(type, base, ref);
+	p.screenpos.x = pos.x / scale[0];
+	p.screenpos.y = pos.y / scale[1];
+	p.screenpos.z = pos.z;
+	m_projected.push_back(p);
+}
+
+// SystemBody can't be inaccessible
+void SystemView::BodyInaccessible(Body *b)
+{
+	if (m_selectedObject.type == Projectable::OBJECT && m_selectedObject.base != Projectable::SYSTEMBODY && m_selectedObject.ref.body == b)
+		ResetViewpoint();
+}
+
+void SystemView::SetVisibility(std::string param)
+{
+	if (param == "RESET_VIEW")
+		ResetViewpoint();
+	else if (param == "GRID_OFF")
+		m_gridDrawing = GridDrawing::OFF;
+	else if (param == "GRID_ON")
+		m_gridDrawing = GridDrawing::GRID;
+	else if (param == "GRID_AND_LEGS")
+		m_gridDrawing = GridDrawing::GRID_AND_LEGS;
+	else if (param == "LAG_OFF")
+		m_showL4L5 = LAG_OFF;
+	else if (param == "LAG_ICON")
+		m_showL4L5 = LAG_ICON;
+	else if (param == "LAG_ICONTEXT")
+		m_showL4L5 = LAG_ICONTEXT;
+	else if (param == "SHIPS_OFF")
+		m_shipDrawing = OFF;
+	else if (param == "SHIPS_ON")
+		m_shipDrawing = BOXES;
+	else if (param == "SHIPS_ORBITS")
+		m_shipDrawing = ORBITS;
+	else if (param == "ZOOM_ON")
+		m_zoomView = true;
+	else if (param == "ROTATE_ON")
+		m_rotateView = true;
+	else if (param == "MOUSE_UP")
+	{
+		m_rotateView = false;
+		m_zoomView = false;
+	}
```

Comment:
Please definitely separate these into their own functions taking boolean arguments. We definitely don't want to be comparing strings if all we're doing is toggling booleans, especially with any regularity!

Reply:
Done! See d9ac37975b2f4161bc1350600ca22d3fe5cedf98